src/control/server/config/server_test.go

```diff
@@ -1276,3 +1274,331 @@ func TestServerConfig_SaveActiveConfig(t *testing.T) {
+		affSrcSet   []EngineAffinityFn
+		expErr      error
+		expDetected uint
+	}{
+		"first source misses; second hits": {
+			cfg: engine.MockConfig(),
+			affSrcSet: []EngineAffinityFn{
+				genAffFn(0, ErrNoAffinityDetected),
+				genAffFn(1, nil),
+			},
+			expDetected: 1,
+		},
+		"first source hits": {
+			cfg: engine.MockConfig(),
+			affSrcSet: []EngineAffinityFn{
+				genAffFn(1, nil),
+				genAffFn(2, nil),
+			},
+			expDetected: 1,
+		},
+		"first source errors": {
+			cfg: engine.MockConfig(),
+			affSrcSet: []EngineAffinityFn{
+				genAffFn(1, errors.New("fatal")),
+				genAffFn(2, nil),
+			},
+			expErr: errors.New("fatal"),
+		},
+		"no sources hit": {
+			cfg: engine.MockConfig(),
+			affSrcSet: []EngineAffinityFn{
+				genAffFn(1, ErrNoAffinityDetected),
+				genAffFn(2, ErrNoAffinityDetected),
+			},
+			expErr: ErrNoAffinityDetected,
+		},
+		"no sources defined": {
+			cfg:    engine.MockConfig(),
+			expErr: ErrNoAffinityDetected,
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			log, buf := logging.NewTestLogger(t.Name())
+			defer ShowBufferOnFailure(t, buf)
+
+			detected, err := detectEngineAffinity(log, tc.cfg, tc.affSrcSet...)
+			CmpErr(t, tc.expErr, err)
+			if tc.expErr != nil {
+				return
+			}
+
+			AssertEqual(t, tc.expDetected, detected,
+				"unexpected detected numa node")
+		})
+	}
+}
+
+func TestConfig_setEngineAffinity(t *testing.T) {
+	for name, tc := range map[string]struct {
+		cfg     *engine.Config
+		setNUMA uint
+		expErr  error
+		expNUMA uint
+	}{
+		"pinned_numa_node set in config overrides detected affinity": {
+			cfg: engine.MockConfig().
+				WithPinnedNumaNode(2).
+				WithFabricInterface("ib1").
+				WithFabricProvider("ofi+verbs"),
+			setNUMA: 1,
+			expNUMA: 2,
+		},
+		"pinned_numa_node not set in config; detected affinity used": {
+			cfg: engine.MockConfig().
+				WithFabricInterface("ib1").
+				WithFabricProvider("ofi+verbs"),
+			setNUMA: 1,
+			expNUMA: 1,
+		},
+		"pinned_numa_node and first_core set": {
+			cfg: engine.MockConfig().
+				WithPinnedNumaNode(2).
+				WithServiceThreadCore(1).
+				WithFabricInterface("ib1").
+				WithFabricProvider("ofi+verbs"),
+			expErr: errors.New("cannot set both"),
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			log, buf := logging.NewTestLogger(t.Name())
+			defer ShowBufferOnFailure(t, buf)
+
+			err := setEngineAffinity(log, tc.cfg, tc.setNUMA)
+			CmpErr(t, tc.expErr, err)
+			if tc.expErr != nil {
+				return
+			}
+
+			AssertEqual(t, tc.expNUMA, *tc.cfg.PinnedNumaNode,
+				"unexpected pinned numa node")
+			AssertEqual(t, tc.expNUMA, tc.cfg.Fabric.NumaNodeIndex,
+				"unexpected numa node in fabric config")
+			AssertEqual(t, tc.expNUMA, tc.cfg.Storage.NumaNodeIndex,
+				"unexpected numa node in storage config")
+		})
+	}
+}
+
+func TestConfig_SetEngineAffinities(t *testing.T) {
+	baseSrvCfg := func() *Server {
+		return DefaultServer()
+	}
+	genAffFn := func(iface string, node uint) EngineAffinityFn {
```

Comment:
Yeah, this is a good point. I added this test before I'd refactored to pull responsibility for affinity detection out of this package. I was thinking that we might want to add more test cases to handle pmem affinity here, but I think you're right that it would be best to keep this package agnostic about those details and just make the test affinity sources completely generic. Unless you feel strongly though, I'd suggest doing this in the follow-on where the code is added to detect pmem affinity.